template-src/com/amazon/corretto/crypto/provider/TemplateHashSpi.java

```diff
@@ -45,7 +45,7 @@ public final class TemplateHashSpi extends MessageDigestSpi implements Cloneable
      * @param buf Input buffer
      */
     // NOTE: This method trusts that all of the array lengths and bufLen are sane.
-    static native void fastDigest(byte[] digest, byte[] buf, int bufLen);
+    static native void fastDigest(byte[] digest, byte[] buf, int bufOffset, int bufLen);
```

Comment:
Can you explain why supporting arbitrary offsets was necessary? I don't see the code path where that ever becomes non-0. The garbage seems to be generated entirely by the length checking logic.

https://github.com/corretto/amazon-corretto-crypto-provider/blob/main/src/com/amazon/corretto/crypto/provider/InputBuffer.java#L299

Reply:
It is not necessary, but the idea to keep it arbitrary is from impl of singlePass which is called by InputBuffer:
```
        if (offset != 0 || length != src.length) {
            src = Arrays.copyOf(src, length);
            offset = 0;
        }
```

Although the code above has bug, correct code should be `src = Arrays.copyOfRange(src, offset, offset + length);
`

Reply:
Ack, thanks for explaining. A reduction in the code footprint, removing buggy logic, and a %-increase in throughput sounds like a win-win-win to me.

Reply:
Thanks!